KAFKA-8841: reduce overhead of ReplicaManager.updateFollowerFetchState

lbradstreet opened the pull request:
This PR makes two changes to code in the ReplicaManager.updateFollowerFetchState path, which is in the hot path for follower fetches. Although calling ReplicaManager.updateFollowerFetch state is inexpensive on its own, it is called once for each partition every time a follower fetch occurs.

1. updateFollowerFetchState no longer calls maybeExpandIsr when the follower is already in the ISR. This avoid repeated expansion checks. 
2. Partition.maybeIncrementLeaderHW is also in the hot path for ReplicaManager.updateFollowerFetchState. Partition.maybeIncrementLeaderHW calls Partition.remoteReplicas four times each iteration, and it performs a toSet conversion. maybeIncrementLeaderHW now avoids generating any intermediate collections when updating the HWM.

**Benchmark results for Partition.updateFollowerFetchState on a r5.xlarge:**
Old:
```
  1288.633 ±(99.9%) 1.170 ns/op [Average]
  (min, avg, max) = (1287.343, 1288.633, 1290.398), stdev = 1.037
  CI (99.9%): [1287.463, 1289.802] (assumes normal distribution)
```

New (when follower fetch offset is updated):
```
  261.727 ±(99.9%) 0.122 ns/op [Average]
  (min, avg, max) = (261.565, 261.727, 261.937), stdev = 0.114
  CI (99.9%): [261.605, 261.848] (assumes normal distribution)
```

New (when follower fetch offset is the same):
```
  68.484 ±(99.9%) 0.025 ns/op [Average]
  (min, avg, max) = (68.446, 68.484, 68.520), stdev = 0.023
  CI (99.9%): [68.460, 68.509] (assumes normal distribution)
```

lbradstreet committed:
Improve maybeExpandIsr performance

lbradstreet committed:
Improve comment

lbradstreet commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -781,7 +789,7 @@ class Partition(val topicPartition: TopicPartition,
     val logStartOffsets = remoteReplicas.collect {
       case replica if metadataCache.getAliveBroker(replica.brokerId).nonEmpty => replica.logStartOffset
-    } + localLogOrException.logStartOffset
+    }.toSet + localLogOrException.logStartOffset
```
We should look at whether we can avoid the set generation here.

lbradstreet commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -763,10 +769,12 @@ class Partition(val topicPartition: TopicPartition,
-        val localLogInfo = (localBrokerId, localLogOrException.logEndOffsetMetadata)
-        trace(s"Skipping update high watermark since new hw $newHighWatermark is not larger than old value. " +
-          s"All current LEOs are ${(replicaInfo + localLogInfo).map(logEndOffsetString)}")
+        if (isTraceEnabled) {
```
Avoid unnecessary collection generation when trace is not enabled.

lbradstreet committed:
Avoid unnecessary toSet in lowWatermarkIfLeader

lbradstreet committed:
Add partition.updateFollowerFetchState benchmark.

lbradstreet committed:
Fix benchmark teardown

lbradstreet committed:
Revert change to number of bench iterations

lbradstreet committed:
Fix jmh benchmarks checkstyle

ijuma commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -749,10 +748,16 @@ class Partition(val topicPartition: TopicPartition,
+        (curTime - replica.lastCaughtUpTimeMs <= replicaLagTimeMaxMs || inSyncReplicaIds.contains(replica.brokerId))) {
+        newHighWatermark = replica.logEndOffsetMetadata
+      }
+    }
```
We should perhaps introduce a method that returns the new high watermark.

ijuma commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -404,8 +404,8 @@ class Partition(val topicPartition: TopicPartition,
 
-  def remoteReplicas: Set[Replica] =
-    remoteReplicasMap.values.toSet
+  def remoteReplicas: Iterable[Replica] =
```
Should we mention that this method is used in the hot path?

ijuma commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -763,10 +768,12 @@ class Partition(val topicPartition: TopicPartition,
+          val localLogInfo = (localBrokerId, localLogOrException.logEndOffsetMetadata)
+          trace(s"Skipping update high watermark since new hw $newHighWatermark is not larger than old value. " +
+            s"All current LEOs are ${(replicaInfo + localLogInfo).map(logEndOffsetString)}")
+        }
```
Good catch.

ijuma commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -779,15 +786,19 @@ class Partition(val topicPartition: TopicPartition,
+      if (metadataCache.getAliveBroker(replica.brokerId).nonEmpty && replica.logStartOffset < lowWaterMark) {
+        lowWaterMark = replica.logStartOffset
+      }
+    }
```
If you want this code to remain this way, you probably need to add some perf related comment (like you did for the other part).

ijuma reviewed (commented):
Thanks for the PR. A few tactical comments inline, but a couple of overall questions:

1. It's interesting that this single collection copy is showing up as significant. `remoteReplicas` is a small collection for RF=3. To be fair, this is a collection copy per partition, but still, I would imagine we have other places where we do a collection copy per partition. Perhaps the hashCode for `Replica` is expensive? Maybe we need to cache that after the first computation?

2. Can we avoid calling `maybeIncrementLeaderHW` altogether in some cases? That would be the biggest win.

lbradstreet commented:
> 1. It's interesting that this single collection copy is showing up as significant. `remoteReplicas` is a small collection for RF=3. To be fair, this is a collection copy per partition, but still, I would imagine we have other places where we do a collection copy per partition. Perhaps the hashCode for `Replica` is expensive? Maybe we need to cache that after the first computation?

@ijuma there are four new sets generated in maybeIncrementLeaderHW as far as I can see.

1 and 2: The toSet in remoteReplicas (1): https://github.com/apache/kafka/pull/7324/files#diff-5450baca03f57b9f2030f93a480e6969L752 and then the map over it (2). The map will generate another set, as it will maintain the collection type of the collection being mapped: https://github.com/apache/kafka/pull/7324/files#diff-5450baca03f57b9f2030f93a480e6969L754
3 and 4: toSet in remoteReplicas (3), plus a map which will generate another set (4): https://github.com/apache/kafka/pull/7324/files#diff-5450baca03f57b9f2030f93a480e6969L766

lbradstreet committed:
Exclude jmh benchmark from spotbugs

lbradstreet committed:
Add perf comment to lowWatermarkIfLeader

lbradstreet committed:
Scala 2.11 deprecation fix

hachikuji commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -749,10 +748,16 @@ class Partition(val topicPartition: TopicPartition,
    * since all callers of this private API acquire that lock
    */
   private def maybeIncrementLeaderHW(leaderLog: Log, curTime: Long = time.milliseconds): Boolean = {
```
Leaving this here for lack of an alternative. I think for low-volume topics, we are actually hitting this logic even when the fetch offset hasn't changed. But if the fetch offset hasn't changed, then there should be no need to expand the ISR or to increment the high watermark. So what I was thinking is that `Replica.updateFetchState` could return a flag indicating whether or not the follower's end offset has changed. If it hasn't, we skip the ISR expansion check. What do you think?

lbradstreet committed:
Avoid maybeExpandIsr and maybeIncrementLeaderHW calls when possible
If the follower end offset has not been updated, then we do not need to
update the leader HWM. If the follower is already in the ISR, we do not
need to check for ISR expansion.

lbradstreet commented:
Here are the updated benchmark results after implementing a strategy to only update leader HWM if the follower fetch offset changed.

Case where the follower fetch offset is updated (previous improvement resulted in 320 ns, now 261 ns):
```
Result "org.apache.kafka.jmh.partition.UpdateFollowerFetchStateBenchmark.updateFollowerFetchStateBench":
  261.727 ±(99.9%) 0.122 ns/op [Average]
  (min, avg, max) = (261.565, 261.727, 261.937), stdev = 0.114
  CI (99.9%): [261.605, 261.848] (assumes normal distribution)
```

Case where follower end offset did not update, and we do not need to update the HWM:
```
Result "org.apache.kafka.jmh.partition.UpdateFollowerFetchStateBenchmark.updateFollowerFetchStateBenchNoChange":
  68.484 ±(99.9%) 0.025 ns/op [Average]
  (min, avg, max) = (68.446, 68.484, 68.520), stdev = 0.023
  CI (99.9%): [68.460, 68.509] (assumes normal distribution)
```

lbradstreet committed:
Remove out of place comment

hachikuji commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -670,10 +683,7 @@ class Partition(val topicPartition: TopicPartition,
           }
-          // check if the HW of the partition can now be incremented
-          // since the replica may already be in the ISR and its LEO has just incremented
-          maybeIncrementLeaderHW(leaderLog, followerFetchTimeMs)
```
I think this is a good change. The intuition is that it is impossible for the high watermark to advance by adding new members to the set. The previous value reflected the smallest end offset among the current replicas in the ISR and the new replica is guaranteed to be at or above that. Whatever the case, the smallest end offset would remain the same.

hachikuji commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -656,7 +669,7 @@ class Partition(val topicPartition: TopicPartition,
+  private def maybeExpandIsr(followerReplica: Replica, followerFetchTimeMs: Long): Unit = {
     inWriteLock(leaderIsrUpdateLock) {
       // check if this replica needs to be added to the ISR
       leaderLogIfLocal match {
```
We can change this to a `foreach`

ijuma replied:
We should also update the scaladoc to remove the result type.

ijuma commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -589,26 +589,39 @@ class Partition(val topicPartition: TopicPartition,
+        // check if the HW of the partition can now be incremented
+        // since the replica may already be in the ISR and its LEO has just incremented
+        val leaderHWIncremented = if (prevFollowerEndOffset != followerReplica.logEndOffset) {
+          leaderLogIfLocal.exists(leaderLog => maybeIncrementLeaderHW(leaderLog, followerFetchTimeMs))
```
We were previously incrementing the hw atomically with the isr expansion. Have we verified that it's ok to change that?

lbradstreet replied:
Hmm, I will look into that. The second problem is that we should probably be taking out a lock in maybeIncrementLeaderHW even if the update does not need to be atomic.

hachikuji replied:
Yeah, I was thinking about this too. We probably need to grab the read side of the `leaderIsrUpdateLock` in `maybeIncrementLeaderHw` to protect from a concurrent ISR change. The second thing is that `Log.maybeIncrementHighWatermark` probably needs to be holding the lock when checking for monotonicity. Currently it only grabs the lock when actually updating the value.

lbradstreet replied:
@hachikuji the changes you suggested seem good to me. Let me know what you think and whether you think I've implemented the locking at the right granularity.

lbradstreet committed:
Document remoteReplicas calls are in hot path

lbradstreet committed:
Fix maybeExpandIsr return documentation, switch to less verbose foreach

lbradstreet committed:
Partition: take out read lock on leaderIsrUpdateLock when incrementaing leader HWM

lbradstreet committed:
Log.maybeIncrementHighWatermark should take out lock to ensure monotonicity check correctness

ijuma reviewed (commented):
@rajinisivaram You had previously made some changes to the locking of this code. Can you please review @lbradstreet's changes?

hachikuji reviewed (approved):
LGTM. Will let @rajinisivaram take a look before merging.

rajinisivaram commented:
@lbradstreet @ijuma I had a quick look at the changes and they look good. Since it has been a long time since I looked at the locking changes, I am not that familiar with the code any more unfortunately. But since @hachikuji has done a thorough review, we should merge it now to get it into the next release.

ijuma commented:
@lbradstreet Can you please update the title to reflect the latest changes (i.e. it's not only about avoiding intermediate collections now).

junrao commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -779,15 +793,21 @@ class Partition(val topicPartition: TopicPartition,
     futureLog match {
       case Some(partitionFutureLog) =>
-        CoreUtils.min(logStartOffsets + partitionFutureLog.logStartOffset, 0L)
+        Math.min(lowWaterMark, partitionFutureLog.logStartOffset)
```
So, it seems that we fixed a bug here? Taking a min with 0 seems wrong. If so, we probably want to reflect that in the jira.

ijuma replied:
@junrao the second parameter for `min` is actually the default if the collection is empty. We should probably rename it to `seqMinOrDefault` to avoid such confusion in the future.

junrao reviewed (commented):
@lbradstreet : Thanks for the PR. LGTM too. A minor comment below.
